internal/voting/events.go

```diff
@@ -13,7 +13,86 @@ import (
 	"github.com/kwilteam/kwil-db/core/types"
 	"github.com/kwilteam/kwil-db/extensions/resolutions"
 	"github.com/kwilteam/kwil-db/internal/sql/versioning"
-	"github.com/kwilteam/kwil-db/internal/voting"
+)
+
+// Life cycle of an event:
+// 1. Events received from an external source is stored in the event store.
+// 2. Block proposer will create resolutions for the events observed.
+// Resolution is a proposal based on an event, to which the network will vote on
+// to get processed and applied.
+// 3. Other voters vote on the existing resolutions if they witnessed the event.
+// 4. Once the resolution meets required threshold votes, the resolution gets
+// approved and processed.
+// 5. Events are deleted from the event store (either in step 2, 3 or 4)
+// depending on the whether you are a proposer or other voter.
+
+const (
+	schemaName = `kwild_events`
+
+	eventStoreVersion = 1
+
+	// eventsTable is the SQL table definition for the events table.
+	// All the events in this table exist in one of the below states.
+	// "received" and "broadcasted" fields are used to track the state of the event:
+	// Newly added and not yet broadcasted:      [broadcasted = false and received = false]
+	// Broadcasted but not included in a block:  [broadcasted = true
+	// Included in a block:                      [received = true]
+	eventsTable = `CREATE TABLE IF NOT EXISTS ` + schemaName + `.events (
+		id BYTEA PRIMARY KEY, -- uuid
+		data BYTEA NOT NULL,
+		event_type TEXT NOT NULL,
+		received BOOLEAN NOT NULL DEFAULT FALSE, -- received is set to true if the network has received the vote for this event
+		broadcasted BOOLEAN NOT NULL DEFAULT FALSE -- broadcasted is set to true if the event has been broadcasted by the validator. It may or may not have been received by the network
+	);`
+	dropEventsTable = `DROP TABLE IF EXISTS ` + schemaName + `.events;`
+
+	insertEventIdempotent = `INSERT INTO ` + schemaName + `.events (id, data, event_type) VALUES ($1, $2, $3) ON CONFLICT DO NOTHING;`
+	deleteEvent           = `DELETE FROM ` + schemaName + `.events WHERE id = $1;`
+
+	deleteEvents = `DELETE FROM ` + schemaName + `.events WHERE id =ANY($1);`
+	getEvents    = `SELECT data, event_type FROM ` + schemaName + `.events;`
+
+	// getNewEvents returns the list of events observed by the validator to which resolutions does not exist.
+	getNewEvents = `SELECT e.data, e.event_type
+	FROM ` + schemaName + `.events AS e
+	LEFT JOIN ` + votingSchemaName + `.resolutions AS r ON e.id = r.id
+	WHERE r.id IS NULL;`
+
+	// eventsToBroadcast returns the list of the resolutionIDs observed by the validator that are not previously broadcasted
+	eventsToBroadcast = `SELECT e.id
+	FROM ` + votingSchemaName + `.resolutions AS r
+	INNER JOIN ` + schemaName + `.events AS e ON r.id = e.id
+	WHERE  e.broadcasted = FALSE;`
```

Comment:
`e.broadcasted` is a `BOOLEAN` so it's more pg idiomatic to say `WHERE NOT e.broadcasted` or `WHERE e.broadcasted`